[#55] Exception 구조화, 공통된 처리 구현 및 에러 코드 추가

Srltas opened the pull request:
## 📌 Summary
Exception을 구조화하고, 공동된 Exception 처리를 구현했습니다.
에러 구분을 위해 에러 코드를 추가해 사용하기로 했습니다.

## 📝 Description
1. 예외 구조화
   - 프로젝트에서 발생할 수 있는 모든 예외는 모두 `RunTogetherException`을 상속하여 관리합니다.
    
   - DB 데이터 조회 실패, API에 대한 권한이 없어 실패하는 경우 발생하는 예외는 별도의 상속 계층을 구성합니다. 
           <img width="300" alt="스크린샷 2024-11-02 오후 12 00 42" src="https://github.com/user-attachments/assets/ca731669-5d69-4c53-97e7-4dddbc437308">

2. 공통된 예외 처리 핸들러 구현
   - `@RestControllerAdvice`를 활용하여 모든 예외를 통합적으로 처리하는 전역 예외 처리 핸들러를 만들었습니다.
   - `ResponseEntityExceptionHandler` 상속 받아 Spring MVC에서 제공하는 예외들은 오버라이딩을 이용해 구현 했습니다.
   - 예외 처리 핸들러에서 Http 상태 코드를 부여하는 방식으로 구현 했습니다.

3. 에러 코드 부여
   - 각 예외 상황에 고유한 에러 코드를 부여하여 관리합니다.
 
4. 표준화된 예외 응답 형식
   - `ErrorResponse` 클래스를 정의하여, 일관된 에러 응답 형식을 반환할 수 있도록 합니다.

## 📚 References
[Kakao Developers](https://developers.kakao.com/docs/latest/ko/rest-api/reference#error-code)
[Exception 전략 블로그 글 1](https://cheese10yun.github.io/spring-guide-exception/#null)
[Exception 전략 블로그 글 2](https://jaehoney.tistory.com/275)

## ✅ Checklist
- [x] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [x] 코딩 스타일 가이드를 준수했습니다.
- [x] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

Srltas committed:
feat: 공통 ErrorCode, Exception 추가

Srltas committed:
feat: 동네 ErrorCode, Exception 추가

Srltas committed:
feat: 사용자 ErrorCode, Exception 추가

Srltas committed:
feat: 공통 예외 처리 구현

- Http 상태 코드를 공통 예외 처리 핸들러에서 결정해서 넣도록 구현

Srltas committed:
refactor: 기존에 사용하던 Exception 제거

Srltas committed:
refactor: User 도메인 Excpetion 변경

Srltas committed:
test: User 도메인 테스트 코드 수정

Srltas committed:
refactor: Service Exception 수정

Srltas committed:
test: 동네 인증 서비스 테스트 코드 수정

Srltas committed:
refactor: 사용자 인증 필터 Exception 수정

Srltas committed:
test: 사용자 인증 필터 테스트 코드 수정

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/exception/BusinessException.java:
```diff
@@ -0,0 +1,10 @@
+
+import com.srltas.runtogether.common.exception.code.ErrorCode;
+
+public abstract class BusinessException extends RunTogetherException {
```
BusinessException 의 용도를 잘 모르겠네요

Srltas replied:
비즈니스 규칙을 위반했을 때 이를 구분하여 처리하고자 만들었습니다.

내 동네 인증과 같은 경우 다음과 같은 비즈니스 규칙 위반이 있을 수 있습니다:
- 내 동네로 등록되지 않은 동네를 인증할 때
- 이미 인증된 동네를 다시 인증할 때
- 허용된 범위 밖에서 동네를 인증할 때

이런 경우 시스템 예외(DB 조회 실패)가 아니라 프로젝트 비즈니스 규칙에 위배되는 상황입니다. 이와 같은 상황에서 발생하는 예외를 모두 `BusinessException`으로 묶어 처리하고자 했습니다.

f-lab-bradley replied:
그럼 이건 Common 이 아니라 application 에 있어야 하는게 아닐까요???????

Srltas replied:
application에 있는 것이 더 맞는거 같습니다!..

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/exception/EntityNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import com.srltas.runtogether.common.exception.code.ErrorCode;
+
+public class EntityNotFoundException extends RunTogetherException {
```
EntityNotFoundException 은 DB 에 데이터가 없을때 발생하는 건가요?
만약 그렇다면 adapters out 에 있으면 안되는걸까요???
여기만 사용될거 같은데

Srltas replied:
common 보단 adapters out에 있는 것이 더 적절해 보입니다.
수정하도록 하겠습니다!

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/exception/handler/BusinessExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+import com.srltas.runtogether.common.exception.BusinessException;
+
+@RestControllerAdvice
+public class BusinessExceptionHandler {
```
다른 Exception Handler 와 분리 한 이유가 있나요?

Srltas replied:
시스템에서 발생할 수 있는 문제(DB 조회 실패)가 아니라 프로젝트 비즈니스 규칙에 위배되는 상황들만 따로 처리하고자 의미상으로 분리해 만들었습니다.
하지만, 다시 생각을 해보니 `RunTogetherExceptionHandler`와 다르게 처리하는 것도 아니고, 분리한 것이 오히려 관리 포인트만 더 늘어나는거 같습니다.
`RunTogetherExceptionHandler`와 합치도록 하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/common/exception/handler/RunTogetherExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+import com.srltas.runtogether.common.exception.UnauthorizedException;
+
+@RestControllerAdvice
+public class RunTogetherExceptionHandler {
```
웹에는 이 에러 말고도 굉장히 많을텐데 다른 에러들은 어떻게 처리 할까요?
스프링에서 기본적으로 제공하는게 있을텐데 그걸 한 번 알아보세요.

그리고 이건 API Response 아닌가요?
그럼 외부의 인터페이스와 연관된 클래스 아닐까요?
common 에 있는게 맞을까요??

Srltas replied:
네 스프링에서 기본적으로 제공하는 예외들을 더 찾아서 처리할 수 있도록 수정하겠습니다!
외부 인터페이스와 연관된 클래스가 맞습니다. common 보다는 adapter에 있는 것이 더 적절해 보입니다. 수정하도록 하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/neighborhood/exception/NeighborhoodNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+
+import com.srltas.runtogether.common.exception.EntityNotFoundException;
+
+public class NeighborhoodNotFoundException extends EntityNotFoundException {
```
이거도 Neighborhood 와 완련된 에러인데 왜 EntityNotFoundException 인가요/??
기준이 뭘까요??

Srltas replied:
프로젝트 비즈니스 규칙에 대한 예외가 아닌 시스템의 문제(DB 조회 실패)라고 생각해 `EntityNotFoundException`을 상속해 구현했습니다.

f-lab-bradley replied:
이게 글로 표현 될 지는 모르겠지만...

application: Neighborhood Id 로 Ropository 를 활용해 DB 조회
mapper: Repository 에서 mapper 를 호출 했을 때 db 에서  enityNotFoundException 발생
Repository: Repository 구현체에서 try 로 mapper 가 던진 EntityNotFoundException 에러를 잡아서 NeighborhoodNotFoundException 던짐

요 플로우는 어때요?

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/user/User.java:
```diff
@@ -35,7 +37,7 @@ public UserNeighborhood verifiedNeighborhood(int neighborhoodId) {
 
 		if (isNull(userNeighborhood)) {
-			throw new CommonException("해당 동네가 등록되지 않았습니다.");
+			throw new NeighborhoodNotRegisteredException();
```
getOrElse() 같은 map 함수를 찾아 보시겠어요??

Srltas replied:
`computeIfAbsent` 메소드를 이용해서 수정해보겠습니다!

f-lab-bradley replied:
넹 제가 ㅋㅋㅋ 리뷰하면서 기억이 안나서.....
ㅈㅅ

Srltas replied:
Map 메소드를 한번 찾아볼 수 있어서 좋았습니다!
피드백 감사합니다👍

Srltas committed:
refactor: computIfAbsent 메소드를 이용해 코드 리펙터링

Srltas committed:
refactor: 비즈니스 예외도 공통 에러 핸들러 처리

Srltas committed:
refactor: adapters out 패키지로 이동

- 외부 API에서 데이터가 없는 경우 발생하는 예외로 adapters out에 위치하는 것이 적절하다 생각해 이동

Srltas committed:
refactor: 예외 핸들러 Adapter로 이동

Srltas committed:
refactor: 스프링 제공 예외 처리 추가

- HttpMessageNotReadableException: 요청 본문이 읽을 수 없는 경우
- HttpRequestMethodNotSupportedException: 지원되지 않는 HTTP 메서드가 호출될 경우
- HttpMediatypeNotSupportedException: 클라이언트가 서버에세 지원하지 않는 미디어 타입을 요청하는 경우
- MissingServletRequestParameterException: 필수 요청 파라미터가 누락된 경우
- TypeMismatchException: 타입 불일치가 발생하는 경우
- BindException: 요청 파라미터 바인딩에 실패하는 경우
- ConstraintViolationException: 제약 조건을 위반하는 경우

Srltas committed:
refactor: 비즈니스 예외 application으로 이동

Srltas committed:
refactor: BusinessException 제거

Srltas committed:
refactor: 불필요한 추상화 제거

- NeighborhoodException, UserException을 제거

Srltas committed:
refactor: Exception 위치 변경

- NotFoundException은 application에서 사용하는 예외로 application으로 위치 변경
- UnauthorizedException은 web 요청에서 권한이 없을 때 발생하는 예외로 adaptor로 위치 변경

Srltas committed:
refactor: 예외 전역 처리 핸들러 위치, 이름 변경

- 프로젝트의 모든 예외를 처리하기 때문에 위치를 common.exception으로 옮김
- 기존의 RunTogetherExceptionHandler는 RuntTogetherException을 상속한 예외만 처리하는 느낌의 이름으로 일반적인 예외도 처리할 수 있는 Handler를 표현하기 위해 GlobalExceptionHandler로 변경

Srltas committed:
refactor: 공통 예외 처리 추가

- 사용자 요청이 잘못된 경우 발생하는 예외들 위주로 추가

Srltas committed:
refactor: GlobalExceptionHandler Adaptor로 위치 변경

- Controller 계층 예외를 처리하기 때문에 Adaptor로 위치 변경

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,104 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+	@ExceptionHandler(EntityNotFoundException.class)
```
여기서 Entity 는 뭔가 어울리지 않는거 같아요.
다른 이름이 좋을것 같은데 이  PR 에서는 일단 넘어가시죠

Srltas replied:
넵!.. 더 적절한 이름을 고민해 보겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/neighborhood/exception/code/NeighborhoodErrorCode.java:
```diff
@@ -0,0 +1,21 @@
+@RequiredArgsConstructor
+public enum NeighborhoodErrorCode implements ErrorCode {
+
+	NEIGHBORHOOD_NOT_FOUND(-301, "해당 동네를 찾을 수 없습니다."),
```
마이너스는 어떤 기준으로 붙이신 거에요????
그냥 궁금궁금

Srltas replied:
[Kakao REST API](https://developers.kakao.com/docs/latest/ko/rest-api/reference#error-code) 문서에서 에러 코드에 `-`를 붙여서 관리하더라고요.
HTTP 상태 코드와 구분을 위해서 이렇게 사용하는 것 같은데 괜찮은거 같아서 한 번 저도 적용해 봤습니다!

f-lab-bradley reviewed (approved):
수고하셨어요
로그도 빨리 붙여 봅시다 우리

Srltas commented:
> 수고하셨어요 로그도 빨리 붙여 봅시다 우리

넵! 오늘 안으로 PR 올리는 것이 목표입니다! 빨리 올리겠습니다.
